File: Partition Equal Subset Sum - GFG/partition-equal-subset-sum.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.util.*;

class GFG{
    public static void main(String args[])throws IOException
    {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(in.readLine());
        while(t-- > 0){
            int N = Integer.parseInt(in.readLine());
            String input_line[] = in.readLine().trim().split("\\s+");
            int arr[] = new int[N];
            for(int i = 0;i < N;i++)
                arr[i] = Integer.parseInt(input_line[i]);
            
            Solution ob = new Solution();
            int x = ob.equalPartition(N, arr);
            if(x == 1)
                System.out.println("YES");
            else
                System.out.println("NO");
        }
    }
}
// } Driver Code Ends


// User function Template for Java

class Solution{
    static int equalPartition(int N, int arr[])
    {
        // code here
        int sum=0;
        for(int i=0;i<N;i++){
            sum+=arr[i];
        }
        
        int w=sum/2;
        int dp[][]=new int[N+1][sum+1];
        
        for(int i=1;i<=N;i++){
            for(int j=1;j<=w;j++){
                if(j>=arr[i-1]){
                    dp[i][j]=Math.max(arr[i-1]+dp[i-1][j-arr[i-1]],dp[i-1][j]);
                }
                else{
                    dp[i][j]=dp[i-1][j];
                }
            }
        }
          int sum1=dp[N][w];
        int sum2=sum-sum1;
        if((sum1-sum2)==0){
            return 1;
        }
        return 0;
    }
}

File: Partition Equal Subset Sum - GFG/partition-equal-subset-sum.js
//{ Driver Code Starts
//Initial Template for javascript

'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(string => {
        return string.trim();
    });
    
    main();    
});

function readLine() {
    return inputString[currentLine++];
}

function main() {
    let t = parseInt(readLine());
    let i = 0;
    for(;i<t;i++)
    {
        let n = parseInt(readLine());
        let arr = new Array(n);
        let input_ar1 = readLine().split(' ').map(x=>parseInt(x));
        for(let i=0;i<n;i++)
            arr[i] = input_ar1[i];
        let obj = new Solution();
        let ok = obj.equalPartition(arr, n);
        if(ok)
            console.log("YES");
        else
            console.log("NO");
    }
}

// } Driver Code Ends


//User function Template for javascript

/**
 * @param {number[]} arr
 * @param {number} n
 * @returns {boolean}
*/

class Solution {
    
    equalPartition(arr, N)
    {
        // Calculate the total sum of elements in the array.
        let sum = 0;
        for(let i = 0;i < N;i++)
            sum = sum + arr[i];
        
        // If sum is odd, return 0 as equal partition is not possible.
        if(sum%2 == 1)
            return 0;
        
        // Calculate the target sum for each subset.
        sum = sum/2;
        
        // Create a 2D array to store the result of subproblems.
        let dp = new Array(N+1);
        for(let i = 0;i<=N;i++){
            dp[i] = new Array(sum+1);
        }
        
        // Initialize the base cases for dynamic programming.
        for(let i= 0;i<dp.length;i++)
            dp[i][0] = true;
        
        for(let i= 1;i<dp.length;i++)
            dp[0][i] = false;
         
        // Use dynamic programming to solve the problem.
        for(let i = 1;i <= N;i++){
            for(let j = 1;j <= sum;j++){
                if(arr[i-1] > j){
                    dp[i][j] = dp[i-1][j];
                }
                else{
                    dp[i][j] = dp[i-1][j] || dp[i-1][j-arr[i-1]];
                }
            }
        }
        
        // Return 1 if equal partition is possible, 0 otherwise.
        return (dp[N][sum]?1:0);
    }
}